src/runtime/string.cpp

```diff
@@ -210,6 +210,8 @@ static PRIMFN(prim_read) {
       runtime.heap.reserve(need_pass);
 
       String *out = String::claim(runtime.heap, size);
+      out->c_str()[size] = 0;
+
       t.seekg(0, t.beg);
       t.read(out->c_str(), out->size());
```

Comment:
You should put this in this claim function

```
String *String::claim(Heap &h, size_t length) {
  return new (h.claim(reserve(length))) String(length);
}
```

turns into

```
String *String::claim(Heap &h, size_t length) {
  auto out = new (h.claim(reserve(length))) String(length);
  out->c_str()[length] = '\0';
  return out;
}
```

like-wise for this function
```
String *String::alloc(Heap &h, size_t length) {
  return new (h.alloc(reserve(length))) String(length);
}
```

The other claim/alloc implementations appear to account for these issues.

Reply:
This is closer to an "API" fix. I personally think the correct API fix is to instead make `out->c_str()` `const` only and to add a `fill(src,len)`/`strcpy(src,len)` method to String that ensures the null byte is always written when externally filling a `String`. All the `String` constructors correctly that upkeep the null byte and we could make fill/strcpy to so as well. The null termination seems a bit unrelated to claim/alloc

Reply:
The contract for string that these uphold is that there is always a null-terminator as long as you don't write outside of the given range. These make it so that the claim/alloc functions always keep that promise.

Reply:
Right, but you shouldn't even be allowed to write outside the given range. Making `c_str` a editable pointer is a clear footgun without any protections. What is the cost of a dedicated `fill` function that makes writing to the raw pointer preferable?

I'm imagining

```
int fill(const char* src, size_t count) {
  assert(count < length);
  memcpy(c_str(), src, count);
  c_str()[count] = 0
}
```

which is also more correct as the null byte always immediately follows the end of the string. If you go the `claim`/`alloc` route and write a string smaller than the reserved space you end up with the opportunity for garbage at the end of the string

Reply:
I'm not against that, but you still need the uninit constructor and that constructor should also maintain the invariant. If you do that you still want my suggested change as well.

Reply:
Which constructor do you mean? I spent time going through them all on Friday and they all maintained the invariant. The only thing that could mess stuff was manually editing the underlying pointer (which `fill` prevents)

Reply:
Spoke off line to clarify, by "constructor" I mean the claim/alloc static member functions. Ashley pointed out that they were just in-place new constructing from an actual constructor and we agreed that the invariant should be maintained there but it wasn't currently being maintained. We slightly disagreed about what the exact invariant should be but we both liked doing a 0 fill so that we could eliminate the complicated case of having uninitialized strings entirely which greatly simplifies the desired invariant.